### PR TITLE
Refactored the fluentd-es addon

### PR DESCRIPTION
**Pull request by crassirostris**
Refactor fluentd-elasticsearch addon:

- Decrease the number of files by moving RBAC-related objects in the same files where they're used
- Move the fluentd configuration out of the image
- Don't use PVC to avoid leaking resources in e2e tests
- Fluentd now ingest docker and kubelet logs that are written to journald
- Disable X-Pack, because it's not free

Fixes https://github.com/kubernetes/kubernetes/issues/41462
Fixes https://github.com/kubernetes/kubernetes/issues/49816
Fixes https://github.com/kubernetes/kubernetes/issues/48973
Fixes https://github.com/kubernetes/kubernetes/issues/49450

@aknuds1 @coffeepac Could you please take a look?

```release-note
Fluentd DaemonSet in the fluentd-elasticsearch addon is configured via ConfigMap and includes journald plugin
Elasticsearch StatefulSet in the fluentd-elasticsearch addon uses local storage instead of PVC by default
```

### CODE REVIEW

**Comment by aknuds1** [comment]
@crassirostris Thanks, will have a look!

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/README.md]: I would say 'just for testing purposes'.

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/README.md]: It's called `emptyDir` rather than `EmptyDir`?

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/README.md]: https://www.elastic.co/products/kibana

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/README.md]: https://www.elastic.co/products/elasticsearch

File: cluster/addons/fluentd-elasticsearch/es-image/config/elasticsearch.yml
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/es-image/config/elasticsearch.yml]: Nice

File: cluster/addons/fluentd-elasticsearch/es-statefulset.yaml
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/es-statefulset.yaml]: 👍

File: cluster/addons/fluentd-elasticsearch/es-statefulset.yaml
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/es-statefulset.yaml]: I thought `elasticsearch-logging` was good to associate the volume with the StatefulSet.

File: cluster/addons/fluentd-elasticsearch/es-statefulset.yaml
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/es-statefulset.yaml]: 👍

File: cluster/addons/fluentd-elasticsearch/fluentd-es-configmap.yaml
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/fluentd-es-configmap.yaml]: Will this still work if no username/password is configured for ES?

File: cluster/addons/fluentd-elasticsearch/fluentd-es-ds.yaml
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/fluentd-es-ds.yaml]: Why the new name?

File: cluster/addons/fluentd-elasticsearch/fluentd-es-ds.yaml
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/fluentd-es-ds.yaml]: How did you determine the new value?

File: cluster/addons/fluentd-elasticsearch/fluentd-es-ds.yaml
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/fluentd-es-ds.yaml]: 👍

File: cluster/addons/fluentd-elasticsearch/fluentd-es-image/Gemfile
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/fluentd-es-image/Gemfile]: 👍

File: cluster/addons/fluentd-elasticsearch/fluentd-es-image/README.md
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/fluentd-es-image/README.md]: https://www.elastic.co/products/elasticsearch

File: cluster/addons/fluentd-elasticsearch/kibana-deployment.yaml
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/kibana-deployment.yaml]: 👍

**Review by aknuds1 — CHANGES_REQUESTED** [review]

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (crassirostris) [inline cluster/addons/fluentd-elasticsearch/README.md]: It's referred as `EmptyDir` in the API docs (e.g. in https://kubernetes.io/docs/api-reference/v1.7/#emptydirvolumesource-v1-core)

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (crassirostris) [inline cluster/addons/fluentd-elasticsearch/README.md]: Ack

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (crassirostris) [inline cluster/addons/fluentd-elasticsearch/README.md]: Ack

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (crassirostris) [inline cluster/addons/fluentd-elasticsearch/README.md]: Yeah, it's funny, I actually noticed that, but it kinda fell in the crack :D

Ack

File: cluster/addons/fluentd-elasticsearch/es-statefulset.yaml
Comment (crassirostris) [inline cluster/addons/fluentd-elasticsearch/es-statefulset.yaml]: Ack

File: cluster/addons/fluentd-elasticsearch/fluentd-es-configmap.yaml
Comment (crassirostris) [inline cluster/addons/fluentd-elasticsearch/fluentd-es-configmap.yaml]: Good catch! Ack

File: cluster/addons/fluentd-elasticsearch/fluentd-es-ds.yaml
Comment (crassirostris) [inline cluster/addons/fluentd-elasticsearch/fluentd-es-ds.yaml]: Dunno, sounds better :)

Ack, bad idea

File: cluster/addons/fluentd-elasticsearch/fluentd-es-ds.yaml
Comment (crassirostris) [inline cluster/addons/fluentd-elasticsearch/fluentd-es-ds.yaml]: Empirically. Note that it's limit, I haven't changed the request. that's just to protect the fluentd from ruby having memory spikes

File: cluster/addons/fluentd-elasticsearch/fluentd-es-image/README.md
Comment (crassirostris) [inline cluster/addons/fluentd-elasticsearch/fluentd-es-image/README.md]: Ack

**Comment by crassirostris** [comment]
@aknuds1 Thanks! I'll address it shortly

File: cluster/addons/fluentd-elasticsearch/fluentd-es-ds.yaml
Comment (coffeepac) [inline cluster/addons/fluentd-elasticsearch/fluentd-es-ds.yaml]: why is this version going backwards?

**Comment by coffeepac** [comment]
this fixes:  #41462

File: cluster/addons/fluentd-elasticsearch/fluentd-es-ds.yaml
Comment (crassirostris) [inline cluster/addons/fluentd-elasticsearch/fluentd-es-ds.yaml]: I decided to move to the semver, but I agree that moving 1.24 -> 0.1.24 and then updating to 0.2.0 is less intuitive than 1.24 -> 1.24.0 -> 2.0.0

I'll change the version, thanks for noticing!

**Comment by crassirostris** [comment]
Thanks for reminding!

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/README.md]: To set up credentials (rather than 'password')?

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/README.md]: "when pods terminate" maybe instead of "with the pod evictions"?

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/README.md]: Maybe say "change the storage to persistent volume claim" instead of "change the storage to the persistent volume claim"?

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/README.md]: I would say "the Fluentd DaemonSet".

File: cluster/addons/fluentd-elasticsearch/fluentd-es-ds.yaml
Comment (crassirostris) [inline cluster/addons/fluentd-elasticsearch/fluentd-es-ds.yaml]: Done

**Review by aknuds1 — CHANGES_REQUESTED** [review]

**Comment by aknuds1** [comment]
@crassirostris Should you document perhaps how the Fluentd configuration must be changed in order to use credentials towards Elasticsearch?

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (crassirostris) [inline cluster/addons/fluentd-elasticsearch/README.md]: Ack

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (crassirostris) [inline cluster/addons/fluentd-elasticsearch/README.md]: Ack

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (crassirostris) [inline cluster/addons/fluentd-elasticsearch/README.md]: Ack

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (crassirostris) [inline cluster/addons/fluentd-elasticsearch/README.md]: Ack

**Review by crassirostris — COMMENTED** [review]
> Should you document perhaps how the Fluentd configuration must be changed in order to use credentials towards Elasticsearch

Ack, will address everything shortly

**Comment by crassirostris** [comment]
@aknuds1 Done, PTAL

**Comment by aknuds1** [comment]
@crassirostris On it

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/README.md]: I think it's better to word it "Since Fluentd talks to the Elasticsearch service inside the cluster, instances on masters won't work, because masters have no kube-proxy. Add a taint on your masters to avoid Fluentd pod scheduling there".

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/README.md]: Is it really necessary to taint masters, since the label `beta.kubernetes.io/fluentd-ds-ready=true` is already used to control the nodes on which Fluentd is scheduled?

File: cluster/addons/fluentd-elasticsearch/fluentd-es-ds.yaml
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/fluentd-es-ds.yaml]: Is there any point in appending the version to the name? We stopped doing this for Elasticsearch itself after all.

**Review by aknuds1 — CHANGES_REQUESTED** [review]

**Review by crassirostris — COMMENTED** [review]
Thanks a lot for your thorough review!

**Review by coffeepac — APPROVED** [review]

**Comment by coffeepac** [comment]
@aknuds1 this looks good to me, but I'll wait for your approval before I put a lgtm label on it

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/README.md]: I think this sounds more correct, as there will be several pods: "when the pods terminate".

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/README.md]: It's not a deployment though, maybe call it 'StatefulSet'?

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/README.md]: 👍

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/README.md]: initContainer, not initContianer

**Review by aknuds1 — CHANGES_REQUESTED** [review]

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (crassirostris) [inline cluster/addons/fluentd-elasticsearch/README.md]: I wanted to say that an EmptyDir volume is erased if the pod it belongs to is restarted

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (crassirostris) [inline cluster/addons/fluentd-elasticsearch/README.md]: Done

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (crassirostris) [inline cluster/addons/fluentd-elasticsearch/README.md]: Wow, thanks for catching this!

**Comment by crassirostris** [comment]
@aknuds1 Thanks! Rephrased Storage paragraph little bit

File: cluster/addons/fluentd-elasticsearch/README.md
Comment (aknuds1) [inline cluster/addons/fluentd-elasticsearch/README.md]: @crassirostris Aha, when reading through again it makes sense in context.

**Review by aknuds1 — APPROVED** [review]

**Pull request merged**
